src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1654 @@
+
+                    // The service gets expired, delete it.
+                    service->DeleteResourcesButRetainName();
+                    earliestExpireTime = OT_MIN(earliestExpireTime, service->GetKeyExpireTime());
+                }
+                else
+                {
+                    earliestExpireTime = OT_MIN(earliestExpireTime, service->GetExpireTime());
+                }
+
+                service = nextService;
+            }
+        }
+
+        host = nextHost;
+    }
+
+    if (earliestExpireTime != now.GetDistantFuture())
+    {
+        if (!mLeaseTimer.IsRunning() || earliestExpireTime <= mLeaseTimer.GetFireTime())
+        {
+            otLogInfoSrp("[server] lease timer is scheduled in %u seconds", (earliestExpireTime - now) / 1000);
+            mLeaseTimer.StartAt(earliestExpireTime, 0);
+        }
+    }
+    else
+    {
+        otLogInfoSrp("[server] lease timer is stopped");
+        mLeaseTimer.Stop();
+    }
+}
+
+void Server::HandleOutstandingUpdatesTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<Server>().HandleOutstandingUpdatesTimer();
+}
+
+void Server::HandleOutstandingUpdatesTimer(void)
+{
+    while (!mOutstandingUpdates.IsEmpty() && mOutstandingUpdates.GetTail()->GetExpireTime() <= TimerMilli::GetNow())
+    {
+        HandleAdvertisingResult(mOutstandingUpdates.GetTail(), OT_ERROR_RESPONSE_TIMEOUT);
+    }
+}
+
+Server::Service *Server::Service::New(Instance &aInstance, const char *aFullName)
+{
+    void *   buf;
+    Service *service = nullptr;
+
+    buf = aInstance.HeapCAlloc(1, sizeof(Service));
+    VerifyOrExit(buf != nullptr);
+
+    service = new (buf) Service(aInstance);
+    if (service->SetFullName(aFullName) != OT_ERROR_NONE)
+    {
+        service->Free();
+        service = nullptr;
+    }
+
+exit:
+    return service;
+}
+
+void Server::Service::Free(void)
+{
+    GetInstance().HeapFree(mFullName);
+    GetInstance().HeapFree(mTxtData);
+    GetInstance().HeapFree(this);
+}
+
+Server::Service::Service(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mFullName(nullptr)
+    , mPriority(0)
+    , mWeight(0)
+    , mPort(0)
+    , mTxtLength(0)
+    , mTxtData(nullptr)
+    , mHost(nullptr)
+    , mNext(nullptr)
+    , mTimeLastUpdate(TimerMilli::GetNow())
+{
+}
+
+otError Server::Service::SetFullName(const char *aFullName)
+{
+    OT_ASSERT(aFullName != nullptr);
+
+    otError error    = OT_ERROR_NONE;
+    char *  nameCopy = static_cast<char *>(GetInstance().HeapCAlloc(1, strlen(aFullName) + 1));
+
+    VerifyOrExit(nameCopy != nullptr, error = OT_ERROR_NO_BUFS);
+    strcpy(nameCopy, aFullName);
+
+    GetInstance().HeapFree(mFullName);
+    mFullName = nameCopy;
+
+exit:
+    return error;
+}
+
+TimeMilli Server::Service::GetExpireTime(void) const
+{
+    OT_ASSERT(!mIsDeleted);
+    OT_ASSERT(!GetHost().IsDeleted());
+
+    return mTimeLastUpdate + GetHost().GetLease() * 1000;
```

Comment:
use existing method to convert s to ms here and the other occurrences.

```suggestion
    return mTimeLastUpdate + Time::SecToMsec(GetHost().GetLease());
```